src/os/freertos/etcpal/os_lock.c

```diff
@@ -0,0 +1,216 @@
+/******************************************************************************
+ * Copyright 2019 ETC Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************
+ * This file is a part of EtcPal. For more information, go to:
+ * https://github.com/ETCLabs/EtcPal
+ ******************************************************************************/
+
+#include "etcpal/lock.h"
+#include <task.h>
+
+/*********************** Private function prototypes *************************/
+
+static void reader_atomic_increment(etcpal_rwlock_t* id);
+static void reader_atomic_decrement(etcpal_rwlock_t* id);
+
+/*************************** Function definitions ****************************/
+
+bool etcpal_mutex_create(etcpal_mutex_t* id)
+{
+  return id ? ((*id = (etcpal_mutex_t)xSemaphoreCreateMutex()) != NULL) : false;
+}
+
+bool etcpal_mutex_take(etcpal_mutex_t* id)
+{
+  if (id && *id)
+  {
+    return (pdTRUE == xSemaphoreTake((SemaphoreHandle_t)*id, portMAX_DELAY));
+  }
+  return false;
+}
+
+bool etcpal_mutex_try_take(etcpal_mutex_t* id)
+{
+  if (id && *id)
+  {
+    return (pdTRUE == xSemaphoreTake((SemaphoreHandle_t)*id, 0));
+  }
+  return false;
+}
+
+void etcpal_mutex_give(etcpal_mutex_t* id)
+{
+  if (id && *id)
+    xSemaphoreGive((SemaphoreHandle_t)*id);
+}
+
```

Comment:
Do you want to add code to allow the semaphore to be given from interrupt context?  Here is code from my PAL layer:

        if (__get_IPSR())
        {
            BaseType_t higherPriorityTaskWoken;

            giveResult = xSemaphoreGiveFromISR(m_semId, &higherPriorityTaskWoken);

            // If higherPriorityTaskWoken is pdTRUE then it indicates to 
            // portYIELD_FROM_ISR that the scheduler should run to allow
            // the new thread to execute when the interrupt handler completes.
            portYIELD_FROM_ISR(higherPriorityTaskWoken);
        }
        else
        {
            // We are not in interrupt context, use standard give
            giveResult = xSemaphoreGive(m_semId);
        }
 
I actually did this for all of my lock code.

Reply:
My only issue with this is that it would introduce a dependency on ARM architecture, which I don't think we have in the FreeRTOS port right now. What if we added another API function like `etcpal_mutex_give_from_isr()` instead?

Reply:
That would work.  Would you be amenable to adding it to the cpp interface as well?

Reply:
Sure!